File: LeetCode Problems/1 to 100/20. Valid Parentheses/20. Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stc;
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
                stc.push(s[i]); continue;
            }
            if(stc.empty()) return false;
            else if (s[i] == ')' && stc.top() != '(') return false;                
            else if (s[i] == '}' && stc.top() != '{') return false;
            else if (s[i] == ']' && stc.top() != '[') return false;
            stc.pop();
        }
        return stc.empty();
    }
};

File: LeetCode Problems/1 to 100/20. Valid Parentheses/20. Valid Parentheses.cs
public class Solution {

    Dictionary<Char, Char> parenthesisDictionary = new Dictionary<Char, Char>() {
        { '{', '}' },
        { '(', ')' },
        { '[', ']' },
    };

    public bool IsValid(string s) {
        Stack<Char> paranthesisChecker = new Stack<Char>();
        for (int i = 0; i < s.Length; i++) {
            if (paranthesisChecker.Count > 0) {
                Char top = paranthesisChecker.Peek();

                if (parenthesisDictionary.ContainsKey(top) && parenthesisDictionary[top] == s[i]) {
                    paranthesisChecker.Pop();
                } else {
                    paranthesisChecker.Push(s[i]);
                }
            } else {
                paranthesisChecker.Push(s[i]);
            }
        }

        return paranthesisChecker.Count == 0 ? true : false;
    }
}
